File: backend/services/zoomService.js
const axios = require('axios');
const logger = require('../utils/logger');

// Zoom API configuration
const ZOOM_API_URL = 'https://api.zoom.us/v2';
const ZOOM_ACCOUNT_ID = process.env.ZOOM_ACCOUNT_ID;
const ZOOM_CLIENT_ID = process.env.ZOOM_CLIENT_ID;
const ZOOM_CLIENT_SECRET = process.env.ZOOM_CLIENT_SECRET;

// In-memory token cache (simple)
let tokenCache = {
    accessToken: null,
    expiresAt: 0
};

async function getZoomAccessToken() {
    if (tokenCache.accessToken && Date.now() < tokenCache.expiresAt) {
        return tokenCache.accessToken;
    }

    try {
        const auth = Buffer.from(`${ZOOM_CLIENT_ID}:${ZOOM_CLIENT_SECRET}`).toString('base64');

        const response = await axios.post('https://zoom.us/oauth/token', null, {
            params: {
                grant_type: 'account_credentials',
                account_id: ZOOM_ACCOUNT_ID
            },
            headers: {
                'Authorization': `Basic ${auth}`
            }
        });

        tokenCache.accessToken = response.data.access_token;
        // Set expiration 5 minutes before actual expiry to be safe
        tokenCache.expiresAt = Date.now() + (response.data.expires_in * 1000) - 300000;

        return tokenCache.accessToken;

    } catch (error) {
        logger.error('Error getting Zoom access token:', error.response?.data || error.message);
        throw new Error('Failed to authenticate with Zoom');
    }
}

async function createMeeting(consultaId, topic, patientName) {
    // Mock mode for development
    if (process.env.NODE_ENV === 'development' || process.env.MOCK_ZOOM === 'true') {
        logger.info('MOCK ZOOM: Creating meeting', { consultaId, topic });
        const mockId = Math.floor(Math.random() * 10000000000).toString();
        return {
            id: mockId,
            topic: topic,
            join_url: `https://zoom.us/j/${mockId}?pwd=mocked`,
            start_url: `https://zoom.us/s/${mockId}?pwd=mocked`,
            password: 'mocked_password',
            settings: {
                join_before_host: true,
                waiting_room: false
            }
        };
    }

    // Real implementation
    try {
        const token = await getZoomAccessToken();

        const meetingData = {
            topic: topic,
            type: 2, // Scheduled meeting
            start_time: new Date().toISOString(),
            duration: 60,
            settings: {
                join_before_host: true,
                waiting_room: false,
                meeting_authentication: false,
                auto_recording: 'none'
            }
        };

        const response = await axios.post(`${ZOOM_API_URL}/users/me/meetings`, meetingData, {
            headers: {
                'Authorization': `Bearer ${token}`,
                'Content-Type': 'application/json'
            }
        });

        return response.data;

    } catch (error) {
        logger.error('Error creating Zoom meeting:', error.response?.data || error.message);
        throw new Error('Failed to create Zoom meeting');
    }
}

async function deleteMeeting(meetingId) {
    if (process.env.NODE_ENV === 'development' || process.env.MOCK_ZOOM === 'true') {
        logger.info('MOCK ZOOM: Deleting meeting', { meetingId });
        return true;
    }

    try {
        const token = await getZoomAccessToken();
        await axios.delete(`${ZOOM_API_URL}/meetings/${meetingId}`, {
            headers: {
                'Authorization': `Bearer ${token}`
            }
        });
        return true;
    } catch (error) {
        logger.warn('Failed to delete meeting from Zoom:', error.message);
        return false;
    }
}

module.exports = {
    createMeeting,
    deleteMeeting
};

File: stixconnect-backend/app/services/zoom_service.py
import requests
import base64
from datetime import datetime, timedelta
from typing import Dict, Optional
from app.core.config import settings

class ZoomService:
    def __init__(self):
        self.account_id = settings.ZOOM_ACCOUNT_ID
        self.client_id = settings.ZOOM_CLIENT_ID
        self.client_secret = settings.ZOOM_CLIENT_SECRET
        self.base_url = "https://api.zoom.us/v2"
        self._access_token = None
        self._token_expiry = None
    
    def _get_access_token(self) -> str:
        if self._access_token and self._token_expiry and datetime.utcnow() < self._token_expiry:
            return self._access_token
        url = f"https://zoom.us/oauth/token?grant_type=account_credentials&account_id={self.account_id}"
        credentials = f"{self.client_id}:{self.client_secret}"
        encoded_credentials = base64.b64encode(credentials.encode()).decode()
        headers = {"Authorization": f"Basic {encoded_credentials}", "Content-Type": "application/x-www-form-urlencoded"}
        response = requests.post(url, headers=headers)
        response.raise_for_status()
        data = response.json()
        self._access_token = data["access_token"]
        self._token_expiry = datetime.utcnow() + timedelta(seconds=data.get("expires_in", 3600) - 300)
        return self._access_token
    
    def create_meeting(self, topic: str, duration: int = 60, timezone: str = "Africa/Luanda", agenda: Optional[str] = None) -> Dict:
        token = self._get_access_token()
        url = f"{self.base_url}/users/me/meetings"
        headers = {"Authorization": f"Bearer {token}", "Content-Type": "application/json"}
        meeting_data = {"topic": topic, "type": 2, "duration": duration, "timezone": timezone, "settings": {"host_video": True, "participant_video": True, "join_before_host": False, "mute_upon_entry": True, "watermark": False, "audio": "both", "auto_recording": "cloud", "waiting_room": False, "approval_type": 2}}
        if agenda:
            meeting_data["agenda"] = agenda
        response = requests.post(url, json=meeting_data, headers=headers)
        response.raise_for_status()
        data = response.json()
        return {"meeting_id": str(data["id"]), "join_url": data["join_url"], "start_url": data["start_url"], "password": data.get("password", "")}

zoom_service = ZoomService()